### ISSUE TITLE
Weird behaviour while logging `CompositionalMetric`

### ISSUE THREAD

**Issue by bojobo** [Lightning-AI/torchmetrics#1833]
## 🐛 Bug

As described in #1756 I'm logging a `CompositionalMetric` as my loss function.
After updating packages to their newest version this week I've been seeing some weird behaviour in calculation of the loss. 
Mostly it leads to a `DataLoader` worker dying.

### To Reproduce

Steps to reproduce: See the code sample of issue #1756
Below I give the code of my `DataModule`

<details>
  <summary>Stack traces and code samples</summary>

```py
[... imports ...]

class BaseDataModule(LightningDataModule):
    def __init__(
            self,
            config
    ):
        super(BaseDataModule, self).__init__()

        self.num_workers = 0 if config["debug"] else 12
        self.pin_memory = not config["debug"]
        self.persistent_workers = not config["debug"]

        self.batch_size = config["batch_size"]
        self.dataset_type = config["type"]

        [... irrelevant variables ...]

    def train_dataloader(self) -> TRAIN_DATALOADERS:
        return self.get_dataloader(self.train_subset, True)

    def val_dataloader(self) -> EVAL_DATALOADERS:
        return self.get_dataloader(self.val_subset)

    def test_dataloader(self) -> EVAL_DATALOADERS:
        return self.get_dataloader(self.test_subset)

    def predict_dataloader(self) -> EVAL_DATALOADERS:
        return self.get_dataloader(self.test_subset)

    def get_dataloader(self, dataset, shuffle=False):
        return DataLoader(
            dataset,
            batch_size=self.batch_size,
            shuffle=shuffle,
            num_workers=self.num_workers,
            pin_memory=self.pin_memory,
            persistent_workers=self.persistent_workers,
        )
```

</details>

I have tried setting `config["debug"]=False` and `config["debug"]=True`. The change didn't fix the bug.
To make this issue easier to read, I've attached the following .log files:

- [01.log](https://github.com/Lightning-AI/torchmetrics/files/11712662/01.log)
- [02.log](https://github.com/Lightning-AI/torchmetrics/files/11712682/02.log)

I think that I have included all relevant code. If not, you can see my whole source code [here (src directory of the project)](https://github.com/bojobo/xmm-superres-denoise/tree/master/src).

### Expected behavior

No errors should occur and the loss should be calculated and logged correctly.

### Environment

- TorchMetrics version (and how you installed TM, e.g. `conda`, `pip`, build from source): torchmetrics: 0.11.4 installed via pip
- Python & PyTorch Version (e.g., 1.0): Python: 3.10, PyTorch: 2.0.1
- Any other relevant information such as OS (e.g., Linux): Running on Ubuntu

**Comment by bojobo** [comment]
I know that there is a fix for #1756, which has to be merged yet, but I'm not sure if they have the same origin.

**Comment by bojobo** [comment]
Hi @SkafteNicki 

- Surprisingly, no. There is no difference between them. Both are `debug=False`. Same errors occur with `debug=True`.
- I haven't tried it, but the error should also occur on randomly generated tensors. I can try to come up with a usable setting without having to download data tomorrow.
- I have some more logs now (also with `debug=False`): 
<img width="950" alt="image" src="https://github.com/Lightning-AI/torchmetrics/assets/17889529/2d4b7169-66d4-4b4f-8f4e-34872b6b6f14">

**Comment by bojobo** [comment]
Done. Steps to start the training: 

1.  Clone my repo at [bojobo/xmm-superres-denoise](https://github.com/bojobo/xmm-superres-denoise)
2. `pip install -r path/to/repo/requirements.txt`
3. Download the following [config.txt](https://github.com/Lightning-AI/torchmetrics/files/11734096/config.txt) and rename it to `.yaml` (I'm not sure if the suffix makes a difference, but just to make sure)
4. Start the fit with `python path/to/train.py fit path/to/config.txt`

I hope that it works. Let me know if something breaks.

**Comment by bojobo** [comment]
@SkafteNicki 

I'll mark this issues as closed, since I believe that something is wrong with my file handling.
The reason why is believe this is twofold:

1. I've started the fit procedure on the `BoringDataset`, which I've created for my last comment, yesterday and it has been running without any issues.
2. At the same time I've also started the fit procedure with the data I'm using on my local machine and it also has been running fine without any issues.

This leads me to believe, that some data on the cluster is either corrupt, not correctly copied to the local space of the worker node, or something else.

I'll reopen this issue if needed.

Thanks for your help!

**Comment by bojobo** [comment]
Well, that didn't take long. I have to reopen this issue, since the same strange behaviour remains when using the "real" dataset.

For whatever reason, the code seems to run on my local machine but a DataLoader worker dies on the cluster.

Since the error occurs while computing my loss (a `CompositionalMetric`), it could be that this issue and issue #1756 are related. @SkafteNicki Could it be that either the missing `self._forward_cache` (PR #1761) leads to this error? Or is my logging (which can be seen [here (lines 91 - 95)](https://github.com/bojobo/xmm-superres-denoise/blob/d293fac20f5e5b8c8867ddfbeaff094049f50749/src/models/model.py#L91)) wrong?

**Comment by SkafteNicki** [comment]
@bojobo alright, do you have any idea what the core difference are between your local setup and the cluster hardware?
Different CPU/GPU? Different version of dependencies (probably not).

It would be very strange to me that worker dying has something to do with the `self._forward_cache` but it does not mean that it can be related.

**Comment by bojobo** [comment]
@SkafteNicki I think the main difference is the used GPU and the corresponding CUDA version.
While the cluster has only RTX 3090 GPUs, my local machine has a RTX 3060.
CUDA version for the cluster is 11.8. I'm not sure what version my local machine has, since I don't have access to it right now.

The dependencies should be the same. Both virtual environments (cluster and local) have been remade on the same day.

**Comment by bojobo** [comment]
Weird. It looks like I've fixed the issues or whatever it was causing it.
The main change was how I log my loss and my validation metrics:

For the train loss I use the forward function of the loss (a `CompositionalMetric`) and reset it after every training epoch.
For the validation loss I use `.update(preds=preds, target=target)` and after the epoch `.compute()` and `.reset()`.
My validation metrics are being logged in the same manner as the validation loss. The main change (which maybe helped solve the issue) here is that I've implemented my own `MetricCollection` which just extends the functionality of `torchmetrics.MetricCollection`.
